File: internal/server/billing.go
package server

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

var legacyPriceNumber = regexp.MustCompile(`[0-9]+(?:\.[0-9]+)?`)

var validCurrencies = map[string]bool{
	"CNY": true,
	"USD": true,
	"HKD": true,
	"EUR": true,
	"JPY": true,
}

var validBillingCycles = map[string]bool{
	"monthly":    true,
	"quarterly":  true,
	"semiannual": true,
	"annual":     true,
	"one_time":   true,
	"custom":     true,
}

func normalizeBilling(price, legacyCycle string, amount float64, currency, cycle string, required bool) (string, float64, string, string, error) {
	price = strings.TrimSpace(price)
	currency = strings.ToUpper(strings.TrimSpace(currency))
	cycle = normalizeBillingCycle(cycle)
	structured := amount > 0 || currency != "" || cycle != ""

	if structured {
		if amount <= 0 || amount > 1_000_000_000 {
			return "", 0, "", "", fmt.Errorf("invalid price_amount")
		}
		if !validCurrencies[currency] {
			return "", 0, "", "", fmt.Errorf("invalid price_currency")
		}
		if !validBillingCycles[cycle] {
			return "", 0, "", "", fmt.Errorf("invalid billing_cycle")
		}
		if price == "" {
			price = formatStructuredPrice(amount, currency)
		}
		return price, amount, currency, cycle, nil
	}

	if price == "" {
		if required {
			return "", 0, "", "", fmt.Errorf("price is required")
		}
		return "", 0, "", "", nil
	}

	parsedAmount, parsedCurrency, parsedCycle := parseLegacyBilling(price, legacyCycle)
	return price, parsedAmount, parsedCurrency, parsedCycle, nil
}

func normalizeBillingCycle(value string) string {
	v := strings.ToLower(strings.TrimSpace(value))
	switch v {
	case "month", "monthly", "月", "月付", "每月":
		return "monthly"
	case "quarter", "quarterly", "季", "季度", "季付":
		return "quarterly"
	case "half-year", "semiannual", "semi-annual", "半年", "半年付":
		return "semiannual"
	case "year", "yearly", "annual", "年", "年付", "每年":
		return "annual"
	case "once", "one-time", "one_time", "一次", "一次性":
		return "one_time"
	case "custom", "自定义":
		return "custom"
	default:
		return v
	}
}

func parseLegacyBilling(price, legacyCycle string) (float64, string, string) {
	compact := strings.ReplaceAll(strings.TrimSpace(price), ",", "")
	upper := strings.ToUpper(compact)
	number := legacyPriceNumber.FindString(compact)
	if number == "" {
		return 0, "", "custom"
	}
	amount, err := strconv.ParseFloat(number, 64)
	if err != nil || amount <= 0 {
		return 0, "", "custom"
	}

	currency := ""
	switch {
	case strings.Contains(upper, "HK$") || strings.Contains(upper, "HKD"):
		currency = "HKD"
	case strings.Contains(upper, "US$") || strings.Contains(upper, "USD") || strings.Contains(compact, "$"):
		currency = "USD"
	case strings.Contains(upper, "JPY") || strings.Contains(upper, "JP¥") || strings.Contains(upper, "JP￥"):
		currency = "JPY"
	case strings.Contains(upper, "CNY") || strings.Contains(upper, "RMB") || strings.Contains(compact, "¥") || strings.Contains(compact, "￥"):
		currency = "CNY"
	case strings.Contains(upper, "EUR") || strings.Contains(compact, "€"):
		currency = "EUR"
	}
	if currency == "" {
		return 0, "", "custom"
	}

	cycleText := strings.ToLower(strings.TrimSpace(legacyCycle + " " + price))
	cycle := "custom"
	switch {
	case strings.Contains(cycleText, "半年") || strings.Contains(cycleText, "semi"):
		cycle = "semiannual"
	case strings.Contains(cycleText, "季度") || strings.Contains(cycleText, "季付") || strings.Contains(cycleText, "quarter"):
		cycle = "quarterly"
	case strings.Contains(cycleText, "月") || strings.Contains(cycleText, "month"):
		cycle = "monthly"
	case strings.Contains(cycleText, "年") || strings.Contains(cycleText, "year") || strings.Contains(cycleText, "annual"):
		cycle = "annual"
	case strings.Contains(cycleText, "一次") || strings.Contains(cycleText, "one-time"):
		cycle = "one_time"
	}
	return amount, currency, cycle
}

func formatStructuredPrice(amount float64, currency string) string {
	formatted := strconv.FormatFloat(amount, 'f', 2, 64)
	formatted = strings.TrimRight(strings.TrimRight(formatted, "0"), ".")
	return currency + " " + formatted
}

func legacyCycleLabel(cycle string) string {
	switch normalizeBillingCycle(cycle) {
	case "monthly":
		return "月"
	case "quarterly":
		return "季"
	case "semiannual":
		return "半年"
	case "annual":
		return "年"
	case "one_time":
		return "一次性"
	default:
		return "自定义"
	}
}

File: web/src/utils/billing.js
const currencies = new Set(['CNY', 'USD', 'HKD', 'EUR', 'JPY'])
const cycles = new Set(['monthly', 'quarterly', 'semiannual', 'annual', 'one_time', 'custom'])

export const normalizeBillingCycle = (value) => {
  const cycle = String(value || '').trim().toLowerCase()
  if (['month', 'monthly', '月', '月付', '每月'].includes(cycle)) return 'monthly'
  if (['quarter', 'quarterly', '季', '季度', '季付'].includes(cycle)) return 'quarterly'
  if (['half-year', 'semiannual', 'semi-annual', '半年', '半年付'].includes(cycle)) return 'semiannual'
  if (['year', 'yearly', 'annual', '年', '年付', '每年'].includes(cycle)) return 'annual'
  if (['once', 'one-time', 'one_time', '一次', '一次性'].includes(cycle)) return 'one_time'
  if (['custom', '自定义'].includes(cycle)) return 'custom'
  return cycle
}

export const parseLegacyBilling = (price, legacyCycle = '') => {
  const raw = String(price || '').trim()
  const compact = raw.replaceAll(',', '')
  const upper = compact.toUpperCase()
  const amount = Number(compact.match(/[0-9]+(?:\.[0-9]+)?/)?.[0] || 0)
  if (!(amount > 0)) return { amount: 0, currency: '', cycle: 'custom', structured: false }

  let currency = ''
  if (upper.includes('HK$') || upper.includes('HKD')) currency = 'HKD'
  else if (upper.includes('US$') || upper.includes('USD') || compact.includes('$')) currency = 'USD'
  else if (upper.includes('JPY') || upper.includes('JP¥') || upper.includes('JP￥')) currency = 'JPY'
  else if (upper.includes('CNY') || upper.includes('RMB') || compact.includes('¥') || compact.includes('￥')) currency = 'CNY'
  else if (upper.includes('EUR') || compact.includes('€')) currency = 'EUR'
  if (!currency) return { amount: 0, currency: '', cycle: 'custom', structured: false }

  const cycleText = `${legacyCycle} ${raw}`.toLowerCase()
  let cycle = 'custom'
  if (cycleText.includes('半年') || cycleText.includes('semi')) cycle = 'semiannual'
  else if (cycleText.includes('季度') || cycleText.includes('季付') || cycleText.includes('quarter')) cycle = 'quarterly'
  else if (cycleText.includes('月') || cycleText.includes('month')) cycle = 'monthly'
  else if (cycleText.includes('年') || cycleText.includes('year') || cycleText.includes('annual')) cycle = 'annual'
  else if (cycleText.includes('一次') || cycleText.includes('one-time')) cycle = 'one_time'
  return { amount, currency, cycle, structured: true }
}

export const billingOf = (item = {}, legacyCycle = '') => {
  const amount = Number(item.price_amount || 0)
  const currency = String(item.price_currency || '').trim().toUpperCase()
  const cycle = normalizeBillingCycle(item.billing_cycle)
  if (amount > 0 && currencies.has(currency) && cycles.has(cycle)) {
    return { amount, currency, cycle, structured: true, legacy: false }
  }
  const parsed = parseLegacyBilling(item.price, legacyCycle || item.cycle)
  return { ...parsed, legacy: parsed.structured }
}

export const cnyValue = (amount, currency, exchange) => {
  const value = Number(amount)
  if (!(value >= 0)) return null
  const code = String(currency || '').toUpperCase()
  if (code === 'CNY') return value
  const rates = exchange?.rates || {}
  const cny = Number(rates.CNY || exchange?.rate || 0)
  const source = Number(rates[code] || (code === 'USD' ? 1 : 0))
  if (!(cny > 0) || !(source > 0)) return null
  return value / source * cny
}

export const monthlyCNY = (billing, exchange) => {
  if (!billing?.structured) return null
  const value = cnyValue(billing.amount, billing.currency, exchange)
  if (value === null) return null
  const divisors = { monthly: 1, quarterly: 3, semiannual: 6, annual: 12 }
  const divisor = divisors[billing.cycle]
  return divisor ? value / divisor : null
}

export const renewalCNY = (billing, exchange) => {
  if (!billing?.structured || !['monthly', 'quarterly', 'semiannual', 'annual'].includes(billing.cycle)) return null
  return cnyValue(billing.amount, billing.currency, exchange)
}

export const formatMoney = (value, currency = 'CNY', locale = 'zh-CN') => {
  const amount = Number(value)
  if (!Number.isFinite(amount)) return '—'
  try {
    return new Intl.NumberFormat(locale, {
      style: 'currency', currency, maximumFractionDigits: amount < 10 ? 2 : 0
    }).format(amount)
  } catch {
    return `${currency} ${amount.toFixed(2)}`
  }
}

export const normalizedPrice = (item, exchange) => {
  const billing = billingOf(item)
  if (!billing.structured) return Number.POSITIVE_INFINITY
  return cnyValue(billing.amount, billing.currency, exchange) ?? Number.POSITIVE_INFINITY
}
